File: backend/Testing/Browser/Page/External/MailinatorPages.cs
using Microsoft.Playwright;

namespace Testing.Browser.Page.External;

public class MailinatorInboxPage : MailInboxPage
{
    public MailinatorInboxPage(IPage page, string mailboxId)
    : base(page, $"https://www.mailinator.com/v4/public/inboxes.jsp?to={mailboxId}", page.Locator($"[id^='row_']").First,
    mailboxId, page.Locator($"[id^='row_']"))
    {
    }

    protected override MailEmailPage GetEmailPage()
    {
        return new MailinatorEmailPage(Page);
    }

    public override async Task<MailInboxPage> Goto(GotoOptions? options = null)
    {
        Url = $"https://www.mailinator.com/v4/public/inboxes.jsp?to={MailboxId}";
        return await base.Goto(options);
    }
}

public class MailinatorEmailPage : MailEmailPage
{
    public MailinatorEmailPage(IPage page) : base(page,
        null, page.FrameLocator("#html_msg_body").Locator("body"))
    {
    }

    public override async Task<string> GetFirstLanguageDepotUrl()
    {
        // Mailinator sometimes swaps links out with its own that and redirect to the original,
        // but the originals are made available in the links tab, which is always in the DOM
        return await Page.Locator("a[href*='jwt=']").First.GetAttributeAsync("href");
    }
}

File: frontend/tests/pages/mailinatorPages.ts
import type { Page } from '@playwright/test';
import { MailEmailPage, MailInboxPage } from './mailPages';

export class MailinatorInboxPage extends MailInboxPage {
  constructor(page: Page, mailboxId: string) {
    super(page,
          page.locator(`[id^='row_']`).first(),
          `https://www.mailinator.com/v4/public/inboxes.jsp?to=${mailboxId}`,
          mailboxId,
          page.locator(`[id^='row_']`));
  }

  override getEmailPage(): MailEmailPage {
    return new MailinatorEmailPage(this.page);
  }

  override async refreshEmails(): Promise<void> {
    return Promise.resolve(); // Mailinator auto-refreshes
  }

  override async goto(options?: {expectRedirect: boolean}): Promise<this> {
    this.url = `https://www.mailinator.com/v4/public/inboxes.jsp?to=${this.mailboxId}`;
    return super.goto(options);
  }
}

export class MailinatorEmailPage extends MailEmailPage {
  constructor(page: Page) {
    super(page, page.frameLocator('#html_msg_body').locator('body'), undefined);
  }

  override getFirstLanguageDepotUrl(): Promise<string | null> {
    // Mailinator sometimes swaps links out with its own that and redirect to the original,
    // but the originals are made available in the links tab, which is always in the DOM
    return this.page.locator('#email_pane').locator(`a[href*='jwt=']`).first().getAttribute('href');
  }
}
